src/internal/pjs/hash.go

```diff
@@ -0,0 +1,109 @@
+package pjs
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"hash"
+	"io"
+	"io/fs"
+
+	"google.golang.org/grpc"
+
+	"github.com/pachyderm/pachyderm/v2/src/storage"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachhash"
+	"github.com/pachyderm/pachyderm/v2/src/internal/tarutil"
+)
+
+// HashFS returns a stable hash of the contents of fileSystem.
+func HashFS(fileSystem fs.FS) ([]byte, error) {
+	var h hasher
+	if err := fs.WalkDir(fileSystem, ".", func(path string, d fs.DirEntry, err error) (retErr error) {
+		if err != nil {
+			return errors.Wrapf(err, "error walking %s", path)
+		}
+		if d.IsDir() {
+			return nil
+		}
+		i, err := d.Info()
+		if err != nil {
+			return errors.Wrapf(err, "info for %s", path)
+		}
+		f, err := fileSystem.Open(path)
+		if err != nil {
+			return errors.Wrapf(err, "open %s", path)
+		}
+		defer errors.Close(&retErr, f, "closing %s", path)
+		if err := h.writeFile("/"+path, i.Size(), f); err != nil {
+			return errors.Wrapf(err, "hashing %s", path)
+		}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "walking testFS")
+	}
+	return h.sum()
+}
+
+// A FilesetReader provides a store.Fileset_ReadFilesetClient for hashing.
+type FilesetReader interface {
+	ReadFileset(context.Context, *storage.ReadFilesetRequest, ...grpc.CallOption) (storage.Fileset_ReadFilesetClient, error)
+}
+
+// HashFileSetID hashes the fileset identified by ID from the FilesetReader fr.
+func HashFilesetID(ctx context.Context, fr FilesetReader, id string) ([]byte, error) {
```

Comment:
[nit]: this sounds like we are hashing the ID itself. I don't think HashFilesetByID sounds much better, so I can't offer a better name.

Reply:
`HashFileset` sounds good.